### PR TITLE
Adding GA Layers from Geoserver using GA WPS Downloads

### PR DESCRIPTION
**Pull request by pmbohm**
Fixes https://github.com/aodn/backlog/issues/947

### CODE REVIEW

**Comment by pmbohm** [comment]
can be tested with

```
					<gmd:onLine>
						<gmd:CI_OnlineResource>
							<gmd:linkage>
								<gmd:URL>http://marine.ga.gov.au/geoserver/wms</gmd:URL>
							</gmd:linkage>
							<gmd:protocol>
								<gco:CharacterString>OGC:WMS-1.1.1-http-get-map</gco:CharacterString>
							</gmd:protocol>
							<gmd:name>
								<gco:CharacterString>marine:bathymetry_150m_all_3857</gco:CharacterString>
							</gmd:name>
							<gmd:description>
								<gco:CharacterString>GA Bathymetry 150m</gco:CharacterString>
							</gmd:description>
						</gmd:CI_OnlineResource>
					</gmd:onLine>
					<gmd:onLine>
						<gmd:CI_OnlineResource>
							<gmd:linkage>
								<gmd:URL>http://nonprod.marine.ga.gov.au/api/products/get-coverage</gmd:URL>
							</gmd:linkage>
							<gmd:protocol>
								<gco:CharacterString>GA:Download</gco:CharacterString>
							</gmd:protocol>
							<gmd:name>
								<gco:CharacterString>imos:argo_primary_profile_core_low_res_good_qc_data</gco:CharacterString>
							</gmd:name>
							<gmd:description>
								<gco:CharacterString>Download subsetted TIFF using WCS</gco:CharacterString>
							</gmd:description>
						</gmd:CI_OnlineResource>
					</gmd:onLine>
```

**Comment by pmbohm** [comment]
@atkinsn Above is a XML fragment for the new GA metadata record
I would expect that the `http://nonprod.marine.ga.gov.au/api/products/get-coverage` URL will change

**Comment by pmbohm** [comment]
@aodn/emii-development-team TIFF Images with correct image size

File: web-app/js/portal/filter/combiner/GaWcsParametersBuilder.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        var newBounds = new OpenLayers.Bounds(wcsParameters.selectedBBOX.split(','));
+        var widthPercentage = newBounds.getWidth() / maxBounds.getWidth();
+        var heightPercentage = newBounds.getHeight() / maxBounds.getHeight();
+        return {
+            width: ((wcsParameters.maxGridEnvelope[0] * widthPercentage) * wcsParameters.scaleFactor).toFixed(0),
+            height: ((wcsParameters.maxGridEnvelope[1] * heightPercentage) * wcsParameters.scaleFactor).toFixed(0)
+        };
+    },
+
+    getClipConfig: function() {
```
Comment (jonescc) [inline web-app/js/portal/filter/combiner/GaWcsParametersBuilder.js]: NB: for prototype only.  Need to use GA service that uses these internally when available or moved to external config.

File: grails-app/services/au/org/emii/portal/JsonService.groovy [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            body = getBody(params)
+            response.success = onResponseSuccess
+        }
+    }
+
+    def getConnection(params) {
+        def conn = new HTTPBuilder(params.server)
+
+        //  If an X-Forwarded-For param was passed - set it as a HTTP parameter
+        if(params["X-Forwarded-For"] != null) {
```
Comment (jonescc) [inline grails-app/services/au/org/emii/portal/JsonService.groovy]: Can we separate normal request parameters from request headers so we don't have to test for individual headers to include?

Comment (jonescc) [reply grails-app/services/au/org/emii/portal/JsonService.groovy]: Also, not really clear what this is for - perhaps explain why its needed in a comment somewhere

File: src/groovy/au/org/emii/portal/wms/GaServer.groovy [hunk 3]
```diff
@@ -0,0 +1,24 @@
+package au.org.emii.portal.wms
```
Comment (jonescc) [inline src/groovy/au/org/emii/portal/wms/GaServer.groovy]: As discussed, this isn't really GA specific - rather its WCS layer specific - worth considering how we handle this when we come to look at removing the filter extension as this filter would likely be common to all WCS layers and optionally time/elevation.

Comment (jonescc) [reply src/groovy/au/org/emii/portal/wms/GaServer.groovy]: In the first case I guess just specifying what it is for each layer would be OK as that's what we do for all our other layers.

File: web-app/js/portal/filter/combiner/GaWcsParametersBuilder.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        });
+
+        return this.buildParameters(geomFilter[0], layerName);
+    },
+
+    buildParameters: function(geomFilter, layerName) {
+
+        var layerConfig = this.getLayerConfig(layerName);
+        layerConfig.selectedBBOX = (geomFilter.value != undefined) ? geomFilter.value.getBounds().toBBOX() : geomFilter.map.getExtent().toBBOX();
+        layerConfig.size = this.getGridEnvelope(layerConfig);
```
Comment (jonescc) [inline web-app/js/portal/filter/combiner/GaWcsParametersBuilder.js]: slectedBBOX and size not really layerConfig?

**Comment by jonescc** [comment]
I'm getting an error going to step 3 after selecting a bbox to download as follows:

```
Uncaught TypeError: Cannot read property 'intersectsBounds' of undefined
    at constructor._checkGeometryIntersectsBounds (SpatialSubsetIntersectTester.js:39)
    at constructor.testSpatialSubsetIntersect (SpatialSubsetIntersectTester.js:17)
    at sb._generateBodyContentForCollection (DownloadPanel.js:125)
    at sb.<anonymous> (DownloadPanel.js:105)
    at Object.each (ext-base-debug.js:458)
    at sb.generateBodyContent (DownloadPanel.js:103)
    at sb.generateContent (DownloadPanel.js:92)
    at sb.<anonymous> (DownloadPanel.js:64)
    at EXTUTIL.Event.fire (ext-all-debug.js:310)
    at sb.fireEvent (ext-all-debug.js:54)
```

**Comment by pmbohm** [comment]
@jonescc which GA layer causes the error?

**Comment by jonescc** [comment]
The one listed above - the bathymetry

**Pull request closed without merging**